cpp/include/cudf/contiguous_split.hpp

```diff
@@ -0,0 +1,198 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cudf/table/table.hpp>
+#include <cudf/types.hpp>
+
+#include <memory>
+#include <vector>
+
+namespace cudf {
+
+/**
+ * @addtogroup column_copy
+ * @{
+ * @file
+ * @brief Table APIs for contiguous_split, pack, unpack, and metadadata
+ */
+
+/**
+ * @brief Column data in a serialized format
+ *
+ * @ingroup copy_split
+ *
+ * Contains data from an array of columns in two contiguous buffers: one on host, which contains
+ * table metadata and one on device which contains the table data.
+ */
+struct packed_columns {
+  packed_columns()
+    : metadata(std::make_unique<std::vector<uint8_t>>()),
+      gpu_data(std::make_unique<rmm::device_buffer>())
+  {
+  }
+
+  /**
+   * @brief Construct a new packed columns object
+   *
+   * @param md Host-side metadata buffer
+   * @param gd Device-side data buffer
+   */
+  packed_columns(std::unique_ptr<std::vector<uint8_t>>&& md,
+                 std::unique_ptr<rmm::device_buffer>&& gd)
+    : metadata(std::move(md)), gpu_data(std::move(gd))
+  {
+  }
+
+  std::unique_ptr<std::vector<uint8_t>> metadata;  ///< Host-side metadata buffer
```

Comment:
Just checking:  what is the intent behind removing the metadata struct and just converting it into a buffer?  Just clarity or some other reason?

Reply:
I've requested to remove it.

The struct `metadata` just wraps around `std::vector` without anything else. There is also not any member functions doing anything on that `std::vector`. So why should we have that?

Reply:
Should we introduce `typedef` or `using` so we can still do `packed_columns::metadata`? If we don't want to see `std::vector<uint8_t>` everywhere...

Reply:
> I've requested to remove it.

The argue I'd make for keeping it is:  it is a return value from `pack_metadata` and by changing it to be a more generic blob of data you're losing a little bit of type safety.   That said, I don't think it's too big of a deal.

Reply:
I think this one could go either way and the next PR introduces a class we can use to build such metadata. Perhaps we can continue on that one. For now, keeping the std::vector<uint8_t> exposed.